### PR TITLE
DSR 1943 DSR-1943 Added ability to register a device and its objects with bluez

### PR DESCRIPTION
**Pull request by dpazj**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -37,6 +37,12 @@ machine.
 
   `./scripts/run-controller.sh`
 
+## Setup 
+
+  We need to install the ble sim dbus config file. Use the following script to install:
+
+  `./scrips/setup.sh`
```
Comment (Tylerg2471) [inline README.md]: typo in scripts

File: scripts/setup.sh [hunk 2]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+cp ./config/blesim.conf /etc/dbus-1/system.d/blesim.conf
```
Comment (Tylerg2471) [inline scripts/setup.sh]: newline at EOF needed

File: src/characteristic.c [hunk 3]
```diff
@@ -5,12 +5,60 @@
...
+static void characteristic_read_value (characteristic_t *characteristic, DBusMessageIter* iter);
+
+DBusObjectPathVTable characteristic_dbus_callbacks = {
+  .unregister_function = characteristic_handle_unregister_device,
+  .message_function = characteristic_handle_dbus_message,
+};
+
+static dbus_property_t characteristic_properties[] =
+{
+  {BLE_PROPERTY_UUID, DBUS_TYPE_STRING_AS_STRING, characteristic_get_uuid}, //s
```
Comment (Tylerg2471) [inline src/characteristic.c]: Comments on this line and the following ones should either be expanded to be meaningful, removed or reworked as a comment block for characteristic_properties to describe whats going on.

File: src/characteristic.h [hunk 4]
```diff
@@ -17,23 +18,26 @@ typedef struct characteristic_t
 {
   char *uuid; //128-bit characteristic UUID.
   char *service_path; //Object path of the GATT service the characteristic belongs to.
```
Comment (Tylerg2471) [inline src/characteristic.h:20]: Is this comment still valid or does object_path fill this role now?

File: src/dbusutils.c [hunk 5]
```diff
@@ -6,16 +6,134 @@
...
+#include <stdlib.h>
+#include <string.h>
+
 #include "dbusutils.h"
 
 bool dbusutils_mainloop_running = false;
 
+void dbusutils_iter_append_string (DBusMessageIter *iter, int type, const char* string)
+{
+  char *tmp = strdup (string);
```
Comment (Tylerg2471) [inline src/dbusutils.c:23]: Question: Worth Null checking string before strdup?

File: src/dbusutils.h [hunk 6]
```diff
@@ -13,12 +13,80 @@
...
+}dbus_property_t;
+
+typedef struct object_flag_t
+{
+  const char* flag_value;
+  unsigned int enabled_bit;
+} object_flag_t;
+
+/**
+ * Appends a string to a dbus iter - trying to appends a string literal causes a segfault
```
Comment (Tylerg2471) [inline src/dbusutils.h]: typo "trying to appends"

File: src/dbusutils.h [hunk 7]
```diff
@@ -13,12 +13,80 @@
...
+ * Populates a dbus message iterator with object data based on a property table
+ * 
+ * @param iter the dbus iter to populate
+ * @param properties_table a NULL terminated properties table
+ * @param object_path the object path
+ * @param interface the interface the object implements
+ * @param object_pointer a pointer to the object that will be passed to the functions in the properties table 
+ **/
+void dbusutils_get_object_data (
+  DBusMessageIter *iter,
```
Comment (Tylerg2471) [inline src/dbusutils.h:53]: Could be worth adding a comment to state what a "iter" is/does

File: src/descriptor.c [hunk 8]
```diff
@@ -5,12 +5,52 @@
...
+static void descriptor_get_value(void *user_data, DBusMessageIter* iter);
+
+static void descriptor_read_value(descriptor_t *descriptor, DBusMessageIter* iter);
+
+DBusObjectPathVTable descriptor_dbus_callbacks = {
+  .unregister_function = descriptor_handle_unregister_device,
+  .message_function = descriptor_handle_dbus_message,
+};
+
+static dbus_property_t descriptor_properties[] =
```
Comment (Tylerg2471) [inline src/descriptor.c:36]: Same as previous comment about removing the comments or updating them as a block. Currently they dont provide meaningful information

**Review by Tylerg2471 — CHANGES_REQUESTED** [review]
Few comments to look at

File: src/characteristic.h (hunk 4 above)
Comment (dpazj) [reply src/characteristic.h:20]: The comment is still valid. A characteristic has its own object_path, and a reference to the service (service_path) which it belongs to.

File: src/dbusutils.c (hunk 5 above)
Comment (dpazj) [reply src/dbusutils.c:23]: Yeah cant hurt, Ill add that in

**Review by Tylerg2471 — APPROVED** [review]
LGTM

**Pull request merged**